### PR TITLE
telemetry upload service

### PR DESCRIPTION
**Pull request by pronoy-debnath**
telemetry service (with systemd)

### CODE REVIEW

File: CDN/telemetry_upload.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    log.info("Device_key[%s] Device_secret[%s] JWT_token[%s]\n" %(device_key, device_secret, tm_jwt))
+
+
+def telemetry_upload_file(filename, jwt, endpoint=tmURL):
+    # Construct a POST request to upload telemetry
+    auth_text = "bearer " + jwt
+    headers = {'Content-Type': 'application/json', 'Content-Encoding': 'gzip', 'Authorization': auth_text}
+
+    fin = open(filename, 'rb')
+    files = {'file': fin}
```
Comment (chetansk) [inline CDN/telemetry_upload.py]: What is the need for this statement ?

File: CDN/telemetry_upload.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        #log.info(' '.join(str(x) for x in tmfile_list))
+
+        #
+        # Upload the first file with existing credential
+        # If we get unauthorized/rate limited error
+        # Handle that
+        #
+        ratelimit_count = 100
+        for filename in tmfile_timesorted_list:
+            status, es_resp_status, es_resp_err, es_resp_errmsg = telemetry_upload_file(filename, tm_jwt, tmURL)
```
Comment (chetansk) [inline CDN/telemetry_upload.py]: The far loop is valid only for files, os.listdir also gives you dir, do you want to check if tmfile_timesorted_list is a dir OR file ?

File: CDN/telemetry_upload.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        # If not connected, just return
+        #
+        netstatus = check_netconnectivity()
+        if not netstatus:
+            err_msg = "Not connected to network..."
+            raise BreakoutException
+        else:
+            log.info("Connected to network...")
+   
+        tmfile_timesorted_list = sorted(tmfile_list, key=os.path.getmtime)
```
Comment (chetansk) [inline CDN/telemetry_upload.py]: sorted gives oldest file first ?  We need to send oldest file first ?

File: CDN/telemetry_upload.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            if es_resp_status == "successful" or es_resp_err == "INVALID_DATA_ERROR":
+                log.info("telemetry upload(%s) status: %s %s" % 
+                        (filename, es_resp_status, es_resp_errmsg))
+                # delete this file
+                os.remove(filename)
+            else:
+                log.info("telemetry upload(%s) status: %d es_status: %s es_err: %s es_errmsg: %s" %
+                        (filename, status, es_resp_status, es_resp_err, es_resp_errmsg))
+                # TODO: If failed due to token expiry, we need to regenerate the token
+                # For now, we are generating the token on every reboot anyway
```
Comment (chetansk) [inline CDN/telemetry_upload.py]: reboot is going to be very rare !!

What is the typical token exire time ?

File: CDN/telemetry_upload.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+            # Ensure we are not rate limited by server
+            if ratelimit_count < 1:
+                break
+            else:
+                ratelimit_count = ratelimit_count - 1
```
Comment (chetansk) [inline CDN/telemetry_upload.py]: Is 100 a good ratelimit ? Ideally this is not rate, but a number ..

File: CDN/telemetry_upload.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        global log_optimization_limit, log_current_value
+        log_current_value  = log_current_value + 1
+
+        if log_current_value == 1:
+            log.error(err_msg)
+        elif log_current_value >= log_optimization_limit:
+            log_current_value  = 0
+            
+    # The below line required for next timer fire
+    threading.Timer(60, telemetry_upload_dir).start()
```
Comment (chetansk) [inline CDN/telemetry_upload.py]: IMO, 60 sec is too fast. This needs to be well designed..

File: CDN/telemetry_upload.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+class BreakoutException(Exception):
+    """ Custom expection """
+    pass
+
+#################
+def logging_init():
+    global log
+    global logfd
+    log = logging.getLogger('TELEMETRY')
+    log.setLevel(logging.DEBUG)
```
[the remark this answers is not in the document]
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: setLevel as logging.INFO

File: CDN/telemetry_upload.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    return (p.returncode)
+
+def jwt_generate(key, secret):
+    payload = { "iss": key }
+    header = { "alg": "HS256", "typ": "JWT"}
+    token = jwt.encode(payload, secret, algorithm=JWT_ALGORITHM, headers=header)
+    return token 
+
+def check_netconnectivity():
+    cmd = "ping -c 2 -W 5 8.8.8.9"
```
[the remark this answers is not in the document]
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: Restored to 8.8.8.8. Was modified for negative testing

File: CDN/telemetry_upload.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    status = run_cmd(cmd)
+    if status == 0:
+        return True
+    # Ping not sucessful, just check with another server
+    cmd = "ping -c 2 -W 5 www.amazon.com"
```
[the remark this answers is not in the document]
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: amazon should be just fine. ekstep.org doesn't responds to ping request.
Yes, if ping fails for both the servers for such issues, it is better to try in the next timer handler.

File: CDN/telemetry_upload.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    #generate a unique device_key
+    global device_key, device_secret, tm_jwt
+
+    #device_key = token_urlsafe(16)
+    device_key = ''.join(random.choice(string.ascii_letters + string.digits) for x in range(16))
+
+    # Construct a POST request to get app key and secret from reqURL
+    payload = {
+            "id": "ekstep.cdn.pinut",
+            "ver": "2.0",
```
[the remark this answers is not in the document]
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: Got this sample code from tarrento developers. This is the playload should be send to server.

File: CDN/telemetry_upload.py (hunk 3 above)
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: Yes, first in first out.

File: CDN/telemetry_upload.py (hunk 2 above)
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: Not needed. This is a well controlled directory created by our code.

File: CDN/telemetry_upload.py (hunk 5 above)
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: Ratelimit logic is in server. We are just taking a precaution not to send too many message in a single shot. Agreed, variable name could have been better.

File: CDN/telemetry_upload.py (hunk 6 above)
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: Any suggestions here?

File: CDN/telemetry_upload.py (hunk 1 above)
Comment (pronoy-debnath) [reply CDN/telemetry_upload.py]: will remove the unused code

**Comment by pronoy-debnath** [comment]
Fix for issue #4

**Pull request merged**